{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9488b800-22ed-430a-843c-bcdb15e50f78",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from collections import Counter\n",
    "\n",
    "import pandas as pd\n",
    "from thefuzz import fuzz\n",
    "from openpyxl import load_workbook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "82693fcd-97e9-48f9-b009-3ce1fb8c9719",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 定义数据文件的路径，使用了 f-string 来构建完整的文件路径。\n",
    "dataset = \"LUAD\"\n",
    "file_id = \"0052ae83-7ae5-470a-a125-5cd94a9fa9e9\"\n",
    "file_name = \"a6a6b9c6-9db7-42b3-a09f-770b7e126fbb.rna_seq.augmented_star_gene_counts.tsv\"\n",
    "sample_path = f\"./TCGA_GDC/{dataset}/samples_info/{file_id}/{file_name}\"\n",
    "\n",
    "# 数据读取和预处理：\n",
    "df = pd.read_csv(sample_path, sep=\"\\t\", skiprows=1)\n",
    "df = df.iloc[4:, :]\n",
    "\n",
    "# 剔除重复基因。删除含有 \"_PAR_Y\" 的基因（这些通常是 Y 染色体上的假常染色体区域的基因）。\n",
    "df = df[~df[\"gene_id\"].str.contains(\"_PAR_Y\")]\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# 创建基因名称到 Ensembl ID 的映射：\n",
    "tmp_dict = {gene_name: ensg_id for gene_name, ensg_id in zip(df[\"gene_name\"], df[\"gene_id\"])}\n",
    "gene_ensg_id_dict = {k: tmp_dict[k].split(\".\")[0] for k in sorted(tmp_dict.keys())}\n",
    "\n",
    "# 将处理后的字典保存为格式化的 JSON 文件：\n",
    "json_str = json.dumps(gene_ensg_id_dict, ensure_ascii=False, indent=2)\n",
    "with open(\"./data/gene_ensg_id_dict.json\", \"w\") as f:\n",
    "    f.write(json_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f0ef46c-a4dd-4ba0-8062-ffe55f995d42",
   "metadata": {},
   "source": [
    "**这段 Python 代码的主要功能是处理一个包含基因信息的 JSON 文件，并检测重复的基因。**\n",
    "\n",
    "1. 首先，使用`with open(\"./data/genes.2023-12-31.json\", \"r\") as f:` 打开名为 \"genes.2023-12-31.json\" 的 JSON 文件，并将文件内容加载到变量 `data` 中。这个文件应该包含一个基因信息的列表，每个基因信息是一个字典。\n",
    "\n",
    "2. 然后，使用 `print(f\"Gene list length: {len(data)}\")` 打印出基因列表的长度，即基因字典的数量。\n",
    "\n",
    "3. 接下来，使用列表推导式 `[dic[\"symbol\"] for dic in data]` 从 `data` 中提取出每个字典的 \"symbol\" 键对应的值，并将结果存储在变量 `genes` 中。\n",
    "\n",
    "4. 使用 `print(Counter(genes).most_common()[:15])` 打印出出现次数最多的前 15 个基因及其出现次数。\n",
    "\n",
    "5. 使用集合推导式 `{dic[\"symbol\"] for dic in data}` 从 `data` 中提取出每个字典的 \"symbol\" 键对应的值，并将结果存储在变量 `genes` 中。这里使用了集合推导式，所以 `genes` 中的元素是唯一的，不会再有重复的基因。\n",
    "\n",
    "6. 使用 `print(f\"Gene set length: {len(genes)}\")` 打印出基因集合的长度。由于集合中的元素是唯一的，所以如果原始数据中有重复的基因，那么这里的基因集合长度会小于基因列表的长度。\n",
    "\n",
    "7. 使用 `[gene for gene in genes]` 将基因集合转换为列表，并将结果存储在变量 `genes` 中。\n",
    "\n",
    "8. 使用 `genes.sort()` 对基因列表进行排序。\n",
    "\n",
    "9. 最后，使用 `with open(\"./data/query_genes.txt\", \"w\") as f:` 打开名为 \"query_genes.txt\" 的文件，并将基因列表中的每个基因写入文件中。如果基因是列表中的最后一个元素，那么只写入基因本身；否则，在基因后面添加一个换行符。\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb9c07bb-fe0f-4a26-bd99-e48a137441ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene list length: 20908\n",
      "[('Y_RNA', 3), ('SFTA3', 2), ('POLR2J3', 2), ('SMIM40', 2), ('PDE11A', 2), ('ACTL10', 2), ('MATR3', 2), ('KCNK3', 1), ('KRTAP5-9', 1), ('SHC4', 1), ('NELL2', 1), ('ACACB', 1), ('RFWD3', 1), ('CALML5', 1), ('PCDH7', 1)]\n",
      "Gene set length: 20900\n"
     ]
    }
   ],
   "source": [
    "# 加载数据。TCGA 项目涉及的基因集共包含 20908 个基因。\n",
    "with open(\"./data/genes.2023-12-31.json\", \"r\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "# 检测重复的基因\n",
    "print(f\"Gene list length: {len(data)}\")             # 打印出 data 的长度，即基因字典的数量。\n",
    "genes = [dic[\"symbol\"] for dic in data]             # 使用列表推导式从 data 中提取出每个字典的 \"symbol\" 键对应的值\n",
    "print(Counter(genes).most_common()[:15])\n",
    "\n",
    "# 某些基因的染色体区段 \"cytoband\" 键对应的值是多个元素的列表（不只是一个），某些基因的字典中多了额外的一个键：\"is_cancer_gene_census\"。\n",
    "genes = {dic[\"symbol\"] for dic in data}             # 使用集合推导式从 data 中提取出每个字典的 \"symbol\" 键对应的值\n",
    "print(f\"Gene set length: {len(genes)}\")             # 20908 - 2 - 6 * 1 = 20900\n",
    "genes = [gene for gene in genes]                    # 将 genes 集合转换为列表\n",
    "genes.sort()                                        # 对 genes 列表进行排序\n",
    "\n",
    "with open(\"./data/query_genes.txt\", \"w\") as f:\n",
    "    f.write(\"\\n\".join(genes))\n",
    "    # for gene in genes:\n",
    "    #     if gene == genes[-1]:\n",
    "    #         f.write(gene)\n",
    "    #     else:\n",
    "    #         f.write(gene + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5cd4b40-b1c6-404f-90d6-8d2c297a1898",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9e009e2b-6780-4f8d-9a5b-4cbe3cf02cb7",
   "metadata": {},
   "source": [
    "**这段代码的主要功能是从一个文本文件和一个 JSON 文件中读取数据，然后生成一个新的字典，并将这个字典保存为一个 JSON 文件。**\n",
    "\n",
    "1. 首先，使用 `with open(\"./data/TCGA_Genes.txt\", \"r\") as f:` 打开名为 \"TCGA_Genes.txt\" 的文件，并将文件中的每一行（即每一个基因）读取出来，存储在集合 `tcga_genes` 中。这里使用了集合推导式，所以 `tcga_genes` 中的元素是唯一的，不会有重复的基因。\n",
    "\n",
    "2. 然后，使用 `with open(\"./data/genes.2023-12-31.json\", \"r\") as f:` 打开名为 \"genes.2023-12-31.json\" 的文件，并使用 `json.load(f)` 将文件中的数据加载出来，存储在变量 `data` 中。\n",
    "\n",
    "3. 接下来，使用字典推导式 `{dic[\"symbol\"]: dic[\"gene_id\"] for dic in data if dic[\"symbol\"] in tcga_genes}` 生成一个新的字典 `gene_to_ensg_id`，其中键是基因的符号，值是基因的 ID。这个字典只包含那些在 `tcga_genes` 中的基因。\n",
    "\n",
    "4. 然后，定义了一个临时字典 `tmp_dict`，并使用 `gene_to_ensg_id.update(tmp_dict)` 将这个临时字典的内容更新到 `gene_to_ensg_id` 中。\n",
    "\n",
    "5. 再次打开 \"TCGA_Genes.txt\" 文件，将文件中的每一行（即每一个基因）读取出来，存储在列表 `tcga_genes` 中。\n",
    "\n",
    "6. 使用字典推导式 `{gene: gene_to_ensg_id[gene] for gene in tcga_genes}` 生成一个新的字典 `save_gene_to_ensg_id`，其中键是基因的名称，值是基因的 ID。\n",
    "\n",
    "7. 最后，使用 `json.dumps(save_gene_to_ensg_id, ensure_ascii=False, indent=2)` 将 `save_gene_to_ensg_id` 转换为 JSON 格式的字符串，并使用 `with open(\"./data/gene_ensg_id_dict.json\", \"w\") as f:` 将这个字符串写入到名为 \"gene_ensg_id_dict.json\" 的文件中。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "303a8859-42cd-4bbf-8fdf-3896f25eba94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 读取 TCGA 基因列表：\n",
    "with open(\"./data/TCGA_Genes.txt\", \"r\") as f:\n",
    "    tcga_genes = {gene for gene in f.read().split(\"\\n\")}\n",
    "\n",
    "# print(tcga_genes[:10], len(tcga_genes))\n",
    "\n",
    "# 加载数据。TCGA 项目涉及的基因集共包含 20908 个基因。\n",
    "with open(\"./data/genes.2023-12-31.json\", \"r\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "# 构建基因到 Ensembl ID 的映射：\n",
    "gene_to_ensg_id = {dic[\"symbol\"]: dic[\"gene_id\"] for dic in data if dic[\"symbol\"] in tcga_genes}\n",
    "# 人工更新一些基因名称及其对应的 Ensembl ID：\n",
    "tmp_dict = {\"SFTA3\": \"ENSG00000229415\", \"POLR2J3\": \"ENSG00000168255\", \"SMIM40\": \"ENSG00000286920\",\n",
    "            \"PDE11A\": \"ENSG00000128655\", \"ACTL10\": \"ENSG00000288649\", \"MATR3\": \"ENSG00000015479\"\n",
    "           }\n",
    "gene_to_ensg_id.update(tmp_dict)\n",
    "\n",
    "with open(\"./data/TCGA_Genes.txt\", \"r\") as f:\n",
    "    tcga_genes = f.read().split(\"\\n\")\n",
    "save_gene_to_ensg_id = {gene: gene_to_ensg_id[gene] for gene in tcga_genes}\n",
    "# 保存映射字典：\n",
    "json_str = json.dumps(save_gene_to_ensg_id, ensure_ascii=False, indent=2)\n",
    "with open(\"./data/gene_ensg_id_dict.json\", \"w\") as f:\n",
    "    f.write(json_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bcef514-29f0-42ce-b0bc-a576522abf36",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7b524fca-3d2f-40e0-b5d7-7a7409c17004",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene ID: 43, Summary Info: Acetylcholinesterase hydrolyzes the neurotransmitter, acetylcholine at neuromuscular junctions and brain cholinergic synapses, and thus terminates signal transmission. It is also found on the red blood cell membranes, where it constitutes the Yt blood group antigen. Acetylcholinesterase exists in multiple molecular forms which possess similar catalytic properties, but differ in their oligomeric assembly and mode of cell attachment to the cell surface. It is encoded by the single ACHE gene, and the structural diversity in the gene products arises from alternative mRNA splicing, and post-translational associations of catalytic and structural subunits. The major form of acetylcholinesterase found in brain, muscle and other tissues is the hydrophilic species, which forms disulfide-linked oligomers with collagenous, or lipid-containing structural subunits. The other, alternatively spliced form, expressed primarily in the erythroid tissues, differs at the C-terminal end, and contains a cleavable hydrophobic peptide with a GPI-anchor site. It associates with the membranes through the phosphoinositide (PI) moieties added post-translationally. AChE activity may constitute a sensitive biomarker of RBC ageing in vivo, and thus, may be of aid in understanding the effects of transfusion[provided by RefSeq, Sep 2019]\n",
      "Gene ID: 2978, Summary Info: This gene encodes an enzyme that plays a role in the recovery of retinal photoreceptors from photobleaching. This enzyme promotes the activity of retinal guanylyl cyclase-1 (GC1) at low calcium concentrations and inhibits GC1 at high calcium concentrations. Mutations in this gene can cause cone dystrophy 3 and code-rod dystrophy 14. provided by RefSeq, Jul 2020]\n",
      "Gene ID: 64135, Summary Info: IFIH1 encodes MDA5 which is an intracellular sensor of viral RNA that triggers the innate immune response. Sensing RNA length and secondary structure, MDA5 binds dsRNA oligonucleotides with a modified DExD/H-box helicase core and a C-terminal domain, thus leading to a proinflammatory response that includes interferons. It has been shown that Coronaviruses (CoVs) as well as various other virus families, are capable of evading the MDA5-dependent interferon response, thus impeding the activation of the innate immune response to infection. MDA5 has also been shown to play an important role in enhancing natural killer cell function in malaria infection. In addition to its protective role in antiviral responses, MDA5 has been implicated in autoimmune and autoinflammatory diseases such as type 1 diabetes, systemic lupus erythematosus, and Aicardi-Goutieres syndrome[provided by RefSeq, Jul 2020]\n"
     ]
    }
   ],
   "source": [
    "with open(\"./data/genes_info.json\", \"r\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "gene_info_dict = {}\n",
    "count_ = 0\n",
    "for gene_id, info in data.items():\n",
    "    if \".[\" in info[\"summary_info\"] or \". [\" in info[\"summary_info\"] or \" [\" in info[\"summary_info\"]:\n",
    "        continue\n",
    "    # if \". [\" not in info[\"summary_info\"] and \" [\" in info[\"summary_info\"]:\n",
    "    #     count_ += 1\n",
    "    #     print(info[\"summary_info\"])\n",
    "    elif \"[provided\" in info[\"summary_info\"] or \" provided \" in info[\"summary_info\"]:\n",
    "        count_ += 1\n",
    "        print(f\"Gene ID: {gene_id}, Summary Info: {info['summary_info']}\")\n",
    "    elif \"NaN\" == info[\"summary_info\"]:\n",
    "        continue\n",
    "    else:\n",
    "        print(\"还有其他情况吗？\")\n",
    "\n",
    "    # 如果计数器 count_ 超过 10，则跳出循环，停止处理更多的基因信息。\n",
    "    if count_ > 10:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2b10d843-da00-4970-b50e-1c3af15536c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This gene encodes an enzyme that plays a role in the recovery of retinal photoreceptors from photobleaching. This enzyme promotes the activity of retinal guanylyl cyclase-1 (GC1) at low calcium concentrations and inhibits GC1 at high calcium concentrations. Mutations in this gene can cause cone dystrophy 3 and code-rod dystrophy 14.\n"
     ]
    }
   ],
   "source": [
    "def check_period(your_str):\n",
    "    \"\"\"确保字符串末尾有句号，如果没有则添加一个句号。\"\"\"\n",
    "    if your_str[-1] == \".\":\n",
    "        return your_str\n",
    "    else:\n",
    "        return your_str + \".\"\n",
    "\n",
    "\n",
    "# 定义摘要文本：\n",
    "summary = \"\"\"This gene encodes an enzyme that plays a role in the recovery of retinal photoreceptors from photobleaching. This enzyme promotes the activity of retinal guanylyl cyclase-1 (GC1) at low calcium concentrations and inhibits GC1 at high calcium concentrations. Mutations in this gene can cause cone dystrophy 3 and code-rod dystrophy 14. provided by RefSeq, Jul 2020]\"\"\"\n",
    "# 查找特定字符串的位置：\n",
    "last_index = summary.rfind(\" provided \")\n",
    "# # 截取字符串并调用 check_period 函数：\n",
    "summary = check_period(summary[:last_index])\n",
    "# 输出处理后的字符串：\n",
    "print(summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "960f11d8-09a4-4700-930e-32d4a21ba0f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "17af871d-dc3a-4a5b-bb79-888a65938ac7",
   "metadata": {},
   "source": [
    "一句话总结：thefuzz 库是一个用于模糊字符串匹配的 Python 库，可以通过使用 fuzz 和 process 方法来进行模糊匹配和提取字符串。\n",
    "\n",
    "要点：\n",
    "\n",
    "- 使用 `pip install thefuzz` 命令安装 thefuzz 库。这个模糊匹配字符串工具通过 Levenshtein Distance 算法（也称为 Edit Distance 算法，来计算字符串之间的最小编辑操作次数。编辑操作包括字符替换、字符插入和字符删除，编辑距离越小，字符串相似度越大。）计算字符串之间的差异。\n",
    "\n",
    "- 使用 `fuzz` 模块中的函数可以计算字符串之间的相似度，如 ratio()、partial_ratio() 和 token_sort_ratio()。这几个 ratio () 函数（方法）得到的结果都是数字，如果需要获得匹配度最高的字符串结果，仍需根据数据类型选择不同函数，并提取结果。\n",
    "\n",
    "- 尽管可以使用此方法量化文本匹配程度，但提取匹配结果并不方便，因此引入了 process 模块。process 模块可以帮助我们在模糊字符串逻辑的帮助下有效地匹配或提取字符串。\n",
    "\n",
    "- 旧版本的 thefuzz 库被称为 fuzzywuzzy，但现在已经不再维护和过时。\n",
    "\n",
    "Github 地址：https://github.com/seatgeek/thefuzz\n",
    "\n",
    "在 Python 中，匹配模糊的字符串：https://cloud.tencent.com/developer/article/2336720\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9dcb099-99d5-4f30-9af1-16617921ec41",
   "metadata": {},
   "source": [
    "使用 `thefuzz` 库进行简单的 Python 模糊字符串匹配测试。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7f284da-6f95-4f88-947f-31d6abc74797",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n",
      "100\n"
     ]
    }
   ],
   "source": [
    "summary = \"\"\"The protein encoded by this gene belongs to the pancreatic ribonuclease family, a subset of the ribonuclease A superfamily. The protein exhibits antimicrobial activity against pathogenic bacteria [provided by RefSeq, Oct 2014]\"\"\"\n",
    "new_summary = \"\"\"The protein encoded by this gene belongs to the pancreatic ribonuclease family, a subset of the ribonuclease A superfamily. The protein exhibits antimicrobial activity against pathogenic bacteria.\"\"\"\n",
    "print(fuzz.partial_ratio(summary, new_summary))\n",
    "\n",
    "summary = \"\"\"This gene encodes a member of the glycosyltransferase family. The specific function of this protein has not been determined. Alternative splicing results in multiple transcript variants of this gene [provided by RefSeq, May 2013]\"\"\"\n",
    "new_summary = \"\"\"This gene encodes a member of the glycosyltransferase family. The specific function of this protein has not been determined. Alternative splicing results in multiple transcript variants of this gene.\"\"\"\n",
    "print(fuzz.partial_ratio(summary, new_summary))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fac96550-bb97-48fd-89e1-858d3935eac0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8e23bccf-2f0f-4878-9e74-8b5041ca5291",
   "metadata": {},
   "source": [
    "要在 Python 中查找字符串最后一次出现位置，可以使用 `rfind ()` 方法。rfind() 方法从字符串末尾开始向前搜索，返回子串第一次出现的索引位置。如果子串不存在，rfind() 将返回 -1。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "88aad718-b1f5-4f75-9feb-3da6c8d014ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------------------------------\n",
      "FuzzyWuzzy：简单易用的字符串模糊匹配工具\n",
      "Gene ID: 53827\n",
      "This gene encodes a member of a family of small membrane proteins that share a 35-amino acid signature sequence domain, beginning with the sequence PFXYD and containing 7 invariant and 6 highly conserved amino acids. The approved human gene nomenclature for the family is FXYD-domain containing ion transport regulator. Mouse FXYD5 has been termed RIC (Related to Ion Channel). FXYD2, also known as the gamma subunit of the Na,K-ATPase, regulates the properties of that enzyme. FXYD1 (phospholemman), FXYD2 (gamma), FXYD3 (MAT-8), FXYD4 (CHIF), and FXYD5 (RIC) have been shown to induce channel activity in experimental expression systems. Transmembrane topology has been established for two family members (FXYD1 and FXYD2), with the N-terminus extracellular and the C-terminus on the cytoplasmic side of the membrane. This gene product, FXYD5, is a glycoprotein that functions in the up-regulation of chemokine production, and it is involved in the reduction of cell adhesion via its ability to down-regulate E-cadherin. It also promotes metastasis, and has been linked to a variety of cancers. Alternative splicing results in multiple transcript variants. [RefSeq curation by Kathleen J. Sweadner, Ph.D., sweadner@helix.mgh.harvard.edu., Sep 2009]\n",
      "This gene encodes a member of a family of small membrane proteins that share a 35-amino acid signature sequence domain, beginning with the sequence PFXYD and containing 7 invariant and 6 highly conserved amino acids. The approved human gene nomenclature for the family is FXYD-domain containing ion transport regulator. Mouse FXYD5 has been termed RIC (Related to Ion Channel). FXYD2, also known as the gamma subunit of the Na,K-ATPase, regulates the properties of that enzyme. FXYD1 (phospholemman), FXYD2 (gamma), FXYD3 (MAT-8), FXYD4 (CHIF), and FXYD5 (RIC) have been shown to induce channel activity in experimental expression systems. Transmembrane topology has been established for two family members (FXYD1 and FXYD2), with the N-terminus extracellular and the C-terminus on the cytoplasmic side of the membrane. This gene product, FXYD5, is a glycoprotein that functions in the up-regulation of chemokine production, and it is involved in the reduction of cell adhesion via its ability to down-regulate E-cadherin. It also promotes metastasis, and has been linked to a variety of cancers. Alternative splicing results in multiple transcript variants.\n",
      "10\n",
      "----------------------------------------------------------------------------------------\n",
      "----------------------------------------------------------------------------------------\n",
      "FuzzyWuzzy：简单易用的字符串模糊匹配工具\n",
      "Gene ID: 53822\n",
      "This reference sequence was derived from multiple replicate ESTs and validated by similar human genomic sequence. This gene encodes a member of a family of small membrane proteins that share a 35-amino acid signature sequence domain, beginning with the sequence PFXYD and containing 7 invariant and 6 highly conserved amino acids. The approved human gene nomenclature for the family is FXYD-domain containing ion transport regulator. Transmembrane topology has been established for two family members (FXYD1 and FXYD2), with the N-terminus extracellular and the C-terminus on the cytoplasmic side of the membrane. FXYD2, also known as the gamma subunit of the Na,K-ATPase, regulates the properties of that enzyme. FXYD1 (phospholemman), FXYD2 (gamma), FXYD3 (MAT-8), FXYD4 (CHIF), and FXYD5 (RIC) have been shown to induce channel activity in experimental expression systems. This gene product, FXYD7, is novel and has not been characterized as a protein. [RefSeq curation by Kathleen J. Sweadner, Ph.D., sweadner@helix.mgh.harvard.edu., Dec 2000]\n",
      "This reference sequence was derived from multiple replicate ESTs and validated by similar human genomic sequence. This gene encodes a member of a family of small membrane proteins that share a 35-amino acid signature sequence domain, beginning with the sequence PFXYD and containing 7 invariant and 6 highly conserved amino acids. The approved human gene nomenclature for the family is FXYD-domain containing ion transport regulator. Transmembrane topology has been established for two family members (FXYD1 and FXYD2), with the N-terminus extracellular and the C-terminus on the cytoplasmic side of the membrane. FXYD2, also known as the gamma subunit of the Na,K-ATPase, regulates the properties of that enzyme. FXYD1 (phospholemman), FXYD2 (gamma), FXYD3 (MAT-8), FXYD4 (CHIF), and FXYD5 (RIC) have been shown to induce channel activity in experimental expression systems. This gene product, FXYD7, is novel and has not been characterized as a protein.\n",
      "10\n",
      "----------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "with open(\"./data/genes_info.json\", \"r\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "def check_period(your_str):   # 检查最后是否有英文句号\n",
    "    if your_str[-1] == \".\":\n",
    "        return your_str\n",
    "    else:\n",
    "        return your_str + \".\"\n",
    "\n",
    "\n",
    "def check_correctness(gene_id, summary, new_summary):\n",
    "    reduce_words = len(summary.split(\" \")) - len(new_summary.split(\" \"))\n",
    "    if reduce_words > 8:\n",
    "        print(\"-\" * 88)\n",
    "        print(\"FuzzyWuzzy：简单易用的字符串模糊匹配工具\")\n",
    "        print(f\"Gene ID: {gene_id}\")\n",
    "        print(summary)\n",
    "        print(new_summary)\n",
    "        print(reduce_words)\n",
    "        print(\"-\" * 88)\n",
    "\n",
    "    if fuzz.partial_ratio(summary, new_summary) < 99.9:\n",
    "        print(\"-\" * 88)\n",
    "        print(\"FuzzyWuzzy：简单易用的字符串模糊匹配工具\")\n",
    "        print(f\"Gene ID: {gene_id}\")\n",
    "        print(summary)\n",
    "        print(new_summary)\n",
    "        print(reduce_words)\n",
    "        print(\"-\" * 88)\n",
    "\n",
    "\n",
    "gene_info_dict = {}\n",
    "for gene_id, info in data.items():\n",
    "    summary = info[\"summary_info\"]\n",
    "    # 字符串模式的校验：\n",
    "    if \".[\" in summary:\n",
    "        last_index = summary.rfind(\".[\")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        continue\n",
    "    elif \". [\" in summary:\n",
    "        last_index = summary.rfind(\". [\")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        continue\n",
    "    elif \" [\" in summary:\n",
    "        last_index = summary.rfind(\" [\")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        continue\n",
    "    elif \"[provided\" in summary:\n",
    "        last_index = summary.rfind(\"[provided\")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        continue\n",
    "    elif \" provided \" in summary:\n",
    "        last_index = summary.rfind(\" provided \")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        continue\n",
    "    elif summary == \"NaN\":\n",
    "        new_summary = summary\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "\n",
    "json_str = json.dumps(gene_info_dict, ensure_ascii=False, indent=2)\n",
    "with open(\"genes_info_clean.json\", \"w\") as f:\n",
    "    f.write(json_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fdae5d3-5e43-4486-9c82-41135b6b5444",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f68b2bd9-52e9-467d-8f57-f332dba18974",
   "metadata": {},
   "source": [
    "从一个 Excel 文件中读取数据，并提取特定列中的超链接，然后将这些超链接添加到原始数据框中作为新的一列。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ec3d0a66-f4d9-4cb7-b7ff-7c6f117f4589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>Samples</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>Papers</th>\n",
       "      <th>Hyperlinks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>40329</td>\n",
       "      <td>844</td>\n",
       "      <td>913</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>14329</td>\n",
       "      <td>124</td>\n",
       "      <td>183</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>3787</td>\n",
       "      <td>58</td>\n",
       "      <td>166</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>4905</td>\n",
       "      <td>332</td>\n",
       "      <td>234</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>63618</td>\n",
       "      <td>1533</td>\n",
       "      <td>1238</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>VHL</td>\n",
       "      <td>38699</td>\n",
       "      <td>2984</td>\n",
       "      <td>903</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>WT1</td>\n",
       "      <td>39330</td>\n",
       "      <td>1493</td>\n",
       "      <td>784</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>XPO1</td>\n",
       "      <td>20593</td>\n",
       "      <td>351</td>\n",
       "      <td>338</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>14801</td>\n",
       "      <td>725</td>\n",
       "      <td>346</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>24627</td>\n",
       "      <td>340</td>\n",
       "      <td>303</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>272 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Genes  Samples  Mutations  Papers  \\\n",
       "0      ABL1    40329        844     913   \n",
       "1     ACVR1    14329        124     183   \n",
       "2    ACVR1B     3787         58     166   \n",
       "3    ACVR2A     4905        332     234   \n",
       "4      AKT1    63618       1533    1238   \n",
       "..      ...      ...        ...     ...   \n",
       "267     VHL    38699       2984     903   \n",
       "268     WT1    39330       1493     784   \n",
       "269    XPO1    20593        351     338   \n",
       "270   ZFHX3    14801        725     346   \n",
       "271   ZRSR2    24627        340     303   \n",
       "\n",
       "                                            Hyperlinks  \n",
       "0    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "1    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "2    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "3    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "4    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "..                                                 ...  \n",
       "267  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "268  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "269  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "270  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "271  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "\n",
       "[272 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# COSMIC - https://cancer.sanger.ac.uk/cosmic/download\n",
    "# Expert Curation of Genes - https://cancer.sanger.ac.uk/cosmic/curation\n",
    "df = pd.read_excel(\"./data/Expert_Curation_of_Genes.xlsx\")\n",
    "wb = load_workbook(\"./data/Expert_Curation_of_Genes.xlsx\")\n",
    "sheet = wb[\"Sheet1\"]\n",
    "\n",
    "hyperlinks = []\n",
    "# 第 1 行是列名，从第 2 行开始有数据。\n",
    "for row in range(2, len(df) + 2):\n",
    "    # 第 1 列的单元格\n",
    "    hyperlink = sheet.cell(row, 1).hyperlink.target\n",
    "    hyperlinks.append(hyperlink)\n",
    "\n",
    "df[\"Hyperlinks\"] = hyperlinks\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9ea8dd19-5c80-4597-a4b5-8f6d6002ce97",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_excel(\"Expert_Curation_of_Genes.xlsx\", index=False, encoding=\"utf-8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "203b4ee6-e7a6-45a4-81cd-c0fedc9527cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0c30b7e-2c24-4d16-86b4-02fc5a2a4b99",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
